BUG: upcasting on reshaping ops #13247

jaehoonhwang opened the pull request:
Only rebasing and fixing the merge conflicts
Original work done by: @jennolsen84
Original branch: https://github.com/jennolsen84/pandas/tree/concatnan
Original PR: #13337 

 - [x] closes #13247
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

jaehoonhwang committed:
BUG: upcasting on reshaping ops #13247 

Only rebasing and fixing the merge conflicts
Original work done by: jennolsen84
Original branch: https://github.com/jennolsen84/pandas/tree/concatnan

jreback commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1899,38 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
         tm.assert_frame_equal(result_no_copy, expected)
+
+
+    def test_concat_no_unnecessary_upcats(self):
```
can u write this in pytest form using parametrize (needs to be a separate function)

jaehoonhwang replied:
Okay, but I need to read up on documentation and it might take some time.

jreback replied:
sure will be something like

```
@parametrize('dtype', np.sctypes('float'))
@parametrize('klass', [Series, DataFrame])
def test_concat_no_unecessary_upcasts(klass, dtype):
     # dtype and klass are defined
```

jreback replied:
FYI only need to do Series/DataFrame/Panel

jaehoonhwang replied:
I need `pytest` module for using `@parametrize`.
Should I create another class for pytest?

jaehoonhwang committed:
Added pytest function

Same testing, but with pytest @parametrize

jaehoonhwang commented:
@jreback Copied the original test and used everything in it, except for `int` testing.

TomAugspurger commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1899,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
+    @parametrize('klass', [Series, DataFrame, Panel])
+    def test_concat_no_unnecessary_upcats_pytest(self, dtype, klass):
+        # GH 13247
+        for pdt in klass:
```
I could be misreading this, but I don't think you need the for loop over `klass` anymore. pytest will handle that for you, since you're doing the parametrize.
I think this will also fail, since this class eventually inherits from `TestCase`, which doesn't play well with parametrized fixtures. I'd try it like this:

```python
import numpy as np
from pandas import Series, DataFrame, Panel
import pandas as pd
import pytest


class TestConcatUpcast(object):

    @pytest.mark.parametrize('pdt', [Series, DataFrame, Panel])
    def test_concat_no_unnecessary_upcats_pytest(self, pdt):
        # GH 13247
        dims = pdt().ndim
        for dt in np.sctypes['float']:
            dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
                   pdt(np.array([np.nan], dtype=dt, ndmin=dims)),
                   pdt(np.array([5], dtype=dt, ndmin=dims))]
            x = pd.concat(dfs)
            assert x.values.dtype == dt

            objs = []
            objs.append(pdt(np.array([1], dtype=np.float32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.float32

            objs = []
            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.int64, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.int64

            objs = []
            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.float64
```

This doesn't parametrize all that well over the `dtype` argument, so probably best to just leavve it in there.

If you want to leave comments on https://github.com/pandas-dev/pandas/issues/15608, would appreciate your insight.

jaehoonhwang replied:
Hmm okay.
I'm trying to set up the test environment, but conda doesn't like me.
I'll try your code after I finish setting it up.

jaehoonhwang committed:
Merge remote-tracking branch 'pandas-dev/master' into Bug13247

jaehoonhwang committed:
Merge remote-tracking branch 'pandas-dev/master'

jaehoonhwang committed:
Merge branch 'master' into Bug13247

jaehoonhwang committed:
Fix: is_float_dtypes and is_numeric_dtype wrong place

Fixed wrong location of is_float_dtypes and is_numeric_dtypes.

jaehoonhwang commented:
There are lots of things that needed to be fixed for this code, so I will keep working on it.

jaehoonhwang committed:
Fix: test_concat.py and internals.py

Everything tests fine, except for pytest.

jaehoonhwang committed:
Working: Except for pytest

Everything tests fine except for the pytest. Waiting for suggestions.

TomAugspurger commented:
> Any guidance for pytest would be appreciated.

you can run individual tests with `pytest pandas -k <regex>`, so `pytest pandas -k test_concat_no_unnecessary_upcast_pytest`.

Let us know if you have other questions.

jreback commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1900,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
         result_no_copy = pd.concat(example_dict, names=['testname'])
         tm.assert_frame_equal(result_no_copy, expected)
+
+    def test_concat_no_unnecessary_upcast(self):
```
this doesn't need to be a class at all, just a function.

jaehoonhwang replied:
Function as function inside of the class or function outside of the class?

jreback replied:
outside

jreback commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1900,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
+                       pdt(np.array([np.nan], dtype=dt, ndmin=dims)),
+                       pdt(np.array([5], dtype=dt, ndmin=dims))]
+                x = pd.concat(dfs)
+                self.assertTrue(x.values.dtype == dt)
```
just do ``assert x.values.dtype == dt``, no need for any reference to ``self``

jaehoonhwang committed:
Merge remote-tracking branch 'pandas-dev/master' into Bug13247

jaehoonhwang committed:
Completed pytest

Pytest completed.
Flake8 —diff completed.

jaehoonhwang commented:
I got pytest working with the code.
@TomAugspurger @jreback Thank you guys for the help.

jaehoonhwang commented:
Hi, do I also need to fix additional errors in `text_index` and `test_reshape`?

jaehoonhwang committed:
Merge remote-tracking branch 'pandas-dev/master' into Bug13247

jreback commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1901,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
+            # not sure what is the best answer here
+            objs = []
+            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
+            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
```
you don't need these tests as you have the ones below

jreback commented on pandas/tests/tools/test_concat.py:
```diff
@@ -1899,3 +1901,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
+
+@pytest.mark.parametrize('pdt', [pd.Series, pd.DataFrame, pd.Panel])
+@pytest.mark.parametrize('dt', np.sctypes['float'])
+def test_concat_no_unnecessary_upcast_pytest(dt, pdt):
```
don't use _pytest in the name, just remove the tests above and use these

jreback commented:
yes everything needs to pass. fix up the tests as indicated and see if you can narrow down where the failures are.

jaehoonhwang committed:
Merge remote-tracking branch 'pandas-dev/master' into Bug13247

jaehoonhwang committed:
Applying request changes removing unnecessary test and renameing

jaehoonhwang commented:
```py
# these are coerced to float unavoidably (as its a list-like to begin)
        df = DataFrame(columns=['A', 'B'])
        df.loc[3] = [6, 7]
    
        exp = DataFrame([[6, 7]], index=[3], columns=['A', 'B'],
                        dtype='float64')
>       tm.assert_frame_equal(df, exp)
```
With this error, I am testing few things and I have few questions.
When I initialize with `df=DataFrame(columns['A','B'])` and do `df.loc[3] = [6,7]`, it is returning me `dtype: object`
If I do `df.loc[3] = [6.0, 7.0]`, A and B are `int64`, but `dtype` of data frame is still `object.

This is not expected behavior correct?

jaehoonhwang committed:
Pass the non-related tests in test_partial and test_reshape

For test_partial, instead of checking for `float`, now it checks for
`object`

For test_reshape, instead of checking for
`pandas.core.internals.SparseBlock` now it checks for
`pandas.core.internals.IntBlock`

jaehoonhwang commented:
@jreback Is this acceptable?

jreback commented:
thanks @jaehoonhwang (and @jennolsen84 )

jaehoonhwang commented:
@jreback Thank you for the help.
